File: js/290_WordPattern.js
/**
 * @param {string} pattern
 * @param {string} str
 * @return {boolean}
 */
var wordPattern = function(pattern, str) {
    var pHash = {};
    var sHash = {};
    var sArray = str.split(' ');
    if(pattern.length != sArray.length) return false;
    for(var i in pattern){
        if( i >= sArray.length) return false;
        if(!pHash[pattern[i]]) pHash[pattern[i]] = i+1;
        if(!sHash[sArray[i]]) sHash[sArray[i]] = i+1;
        if(pHash[pattern[i]] != sHash[sArray[i]]) return false;
    }
    return true;
};

File: cpp/290_WordPattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        std::unordered_map<char,intvar> pHash;
        std::unordered_map<string,int> sHash;
        int sCount = 0;
        for(int i = 0; i < pattern.size(); ++i){
            string tmp;
            if(sCount >= str.length()) return false;
            while(sCount < str.length() && str[sCount] != ' '){
                tmp += str[sCount];
                sCount++;
            }
            if(pHash[pattern[i]] != sHash[tmp]) return false;
            else pHash[pattern[i]] = sHash[tmp] = i+1;
            
            sCount++;
        }
        return sCount == str.length() + 1;
    }
};